synapse/handlers/e2e_keys.py

```diff
@@ -984,56 +980,47 @@ def _get_e2e_cross_signing_verify_key(self, user_id, key_type, from_user_id=None
+            and not self.is_mine(user)
+            # We only get "master" and "self_signing" keys from remote servers
+            and desired_key_type in ["master", "self_signing"]
+        ):
             try:
-                remote_result = yield self.federation.query_client_keys(
-                    user.domain, {"device_keys": {user_id: []}}, timeout=10 * 1000
+                remote_result = yield self.federation.query_user_devices(
+                    user.domain, user_id
                 )
 
-                # Process the result
-                for remote_key_type, remote_user_dict in remote_result.items():
-                    # The key_type variable passed to this function is in the form
-                    # "self_signing","master" etc. whereas the results returned from
-                    # the remote server use "self_signing_keys", "master_keys" etc.
-                    # Remove the "_keys" from the key type
-                    if remote_key_type.endswith("_keys"):
-                        remote_key_type = remote_key_type[:-5]
-
-                    # remote_user_dict is a dictionary in the form of
-                    # {
-                    #   "user_id": {
-                    #     "master_keys": ...
-                    #   },
-                    #   ...
-                    # }
-
-                    # Only extract the keys that pertain to the requested user
-                    key_content_list = remote_user_dict.get(user_id, {}).values()
-
-                    for key_content in key_content_list:
-                        # If the key_type here matches the key we're requesting,
-                        # then this is the key we want to return
-                        if remote_key_type == key_type:
-                            key = key_content
-
-                        # At the same time, save the key to the database for subsequent
-                        # queries
-                        yield self.store.set_e2e_cross_signing_key(
-                            user_id, remote_key_type, key_content
-                        )
-            except (
-                HttpResponseException,
-                NotRetryingDestination,
-                FederationDeniedError,
-                RequestSendFailed,
-            ) as e:
+                # Process each of the retrieved cross-signing keys
+                for key_type in ["master", "self_signing"]:
+                    key_content = remote_result.get(key_type + "_key")
+                    if not key_content:
+                        continue
+
+                    # If this is the desired key type, return it
+                    if key_type == desired_key_type:
+                        key = key_content
+
+                    # At the same time, store this key in the db for
+                    # subsequent queries
+                    yield self.store.set_e2e_cross_signing_key(
+                        user_id, key_type, key_content
+                    )
```

Comment:
do we not have to worry about how this cache will get invalidated when the situation changes in future?

Reply:
Which cache do you mean? `_get_bare_e2e_cross_signing_keys`?

Reply:
I mean the database table itself, which is itself a cache of remote devices.

I'm concerned that the presence of an entry in the database will make other parts of the code believe that the cache for that user/device is up-to-date, which it may not be. In particular: what happens when we *do* start sharing a room with that user: typically I'd expect a device-list-sync to take place, but I'm concerned that the presence of this entry may inhibit that behaviour.

On the other hand, this may be very similar to the behaviour when we *stop* sharing a room with a user: if there is no attempt to purge the database table when that happens, then I guess that implies it is safe to have entries in the table even when we do not share a room with the user.

In short: it might all be fine, it might not, but I think we (you :stuck_out_tongue: ) should have a look around to confirm this.

Reply:
From what I can see, anything that references this table seems to be prompted by remote or local clients requesting device keys.

So I think if anything were to break from this, it would be an assumption the client makes. Which we may only be able to discover by testing?

Reply:
ok yes, I think you're right. I'd somewhat failed to grok that these aren't treated quite the same way as regular devices.